cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/batch/dataset/partitioned/DynamicPartitioningOutputFormat.java

```diff
@@ -0,0 +1,189 @@
+import com.google.common.reflect.TypeToken;
+import org.apache.hadoop.fs.Path;
+import org.apache.hadoop.mapred.InvalidJobConfException;
+import org.apache.hadoop.mapreduce.Job;
+import org.apache.hadoop.mapreduce.JobContext;
+import org.apache.hadoop.mapreduce.RecordWriter;
+import org.apache.hadoop.mapreduce.TaskAttemptContext;
+import org.apache.hadoop.mapreduce.TaskType;
+import org.apache.hadoop.mapreduce.lib.output.FileOutputCommitter;
+import org.apache.hadoop.mapreduce.lib.output.FileOutputFormat;
+import org.apache.hadoop.mapreduce.lib.output.TextOutputFormat;
+import org.apache.hadoop.mapreduce.security.TokenCache;
+import org.apache.hadoop.mapreduce.task.TaskAttemptContextImpl;
+
+import java.io.IOException;
+import java.util.TreeMap;
+
+import static org.apache.hadoop.mapreduce.TaskType.MAP;
+
+/**
+ * This class extends the FileOutputFormat and allows writing dynamically to multiple partitions of a PartitionedFileSet
+ * Dataset.
+ *
+ * @param <K> Type of key
+ * @param <V> Type of value
+ */
+public class DynamicPartitioningOutputFormat<K, V> extends FileOutputFormat<K, V> {
+
+  public static final String HCONF_ATTR_OUTPUT_DATASET = "output.dataset.name";
+
+  // TODO: keep the limit to only FileOutputFormat and FileOutputCommitter?
+  private FileOutputFormat<K, V> fileOutputFormat;
+  private FileOutputCommitter committer;
+
+  /**
+   * Create a composite record writer that can write key/value data to different output files.
+   *
+   * @return a composite record writer
+   * @throws IOException
+   */
+  @Override
+  public RecordWriter<K, V> getRecordWriter(final TaskAttemptContext job) throws IOException {
+    final String outputName = FileOutputFormat.getOutputName(job);
+
+    Class<? extends DynamicPartitioner> className = job.getConfiguration()
+      .getClass(PartitionedFileSetArguments.DYNAMIC_PARTITIONER_CLASS_NAME, null, DynamicPartitioner.class);
+
+    @SuppressWarnings("unchecked")
+    final DynamicPartitioner<K, V> dynamicPartitioner =
+      new InstantiatorFactory(false).get(TypeToken.of(className)).create();
+
+    MapReduceMetrics.TaskType taskType = MapReduceMetrics.TaskType.from(job.getTaskAttemptID().getTaskType());
+    final BasicMapReduceTaskContext<K, V> mrTaskContext = new MapReduceTaskContextProvider(job, taskType).get();
+    String outputDatasetName = job.getConfiguration().get(HCONF_ATTR_OUTPUT_DATASET);
+    PartitionedFileSet outputDataset = mrTaskContext.getDataset(outputDatasetName);
+    final Partitioning partitioning = outputDataset.getPartitioning();
+
+    dynamicPartitioner.initialize(mrTaskContext);
+
+    return new RecordWriter<K, V>() {
+
+      // a cache storing the record writers for different output files.
+      TreeMap<String, RecordWriter<K, V>> recordWriters = new TreeMap<>();
+
+      public void write(K key, V value) throws IOException, InterruptedException {
+        PartitionKey partitionKey = dynamicPartitioner.getPartitionKey(key, value);
```

Comment:
perhaps it would be better to use the partition key as the key in the cache. Otherwise, we create the output path, and the the final path for every single record that is written. That is a lot of overhead in creating the strings.